### PR TITLE
Core: Fix numeric overflow of timestamp nano literal

### PR DESCRIPTION
**Pull request by ebyhr**
The previous logic leads to overflow when we pass timestamp nanos long value.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 1]
```diff
@@ -300,8 +300,7 @@ public <T> Literal<T> to(Type type) {
         case TIMESTAMP:
           return (Literal<T>) new TimestampLiteral(value());
         case TIMESTAMP_NANO:
-          // assume micros and convert to nanos to match the behavior in the timestamp case above
-          return new TimestampLiteral(value()).to(type);
+          return (Literal<T>) new TimestampNanoLiteral(value());
```
Comment (amogh-jahagirdar) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: This change seems correct to me, the previous behavior for this was to assume the value was in microseconds and then pass that through to TimestampLiteral but that can overflow and does not actually represent a nanosecond timestamp!

File: api/src/test/java/org/apache/iceberg/types/TestConversions.java [hunk 2]
```diff
@@ -111,9 +111,9 @@ public void testByteBufferConversions() {
     assertConversion(
         400000000L, TimestampNanoType.withZone(), new byte[] {0, -124, -41, 23, 0, 0, 0, 0});
     assertThat(Literal.of(400000L).to(TimestampNanoType.withoutZone()).toByteBuffer().array())
-        .isEqualTo(new byte[] {0, -124, -41, 23, 0, 0, 0, 0});
+        .isEqualTo(new byte[] {-128, 26, 6, 0, 0, 0, 0, 0});
     assertThat(Literal.of(400000L).to(TimestampNanoType.withZone()).toByteBuffer().array())
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/types/TestConversions.java:115]: I'm a bit confused how the original assertion was passing? Shouldn't have this always been equivalent to `{-128, 26, 6, 0, 0, 0, 0, 0}`?

File: api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java [hunk 3]
```diff
@@ -107,6 +108,28 @@ public void testTimestampMicrosToDateConversion() {
     assertThat(dateOrdinal).isEqualTo(-1);
   }
 
+  @Test
+  public void testTimestampNanoWithLongLiteral() {
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java:112]: could we also add a case for `withZone`?

File: api/src/test/java/org/apache/iceberg/types/TestConversions.java (hunk 2 above)
Comment (ebyhr) [reply api/src/test/java/org/apache/iceberg/types/TestConversions.java:115]: I think the cause is the original logic called `DateTimeUtil.microsToNanos` method which multiples the value by 1000:
https://github.com/apache/iceberg/blob/b9b61b1d72ebb192d5e90453ff7030ece73d2603/api/src/main/java/org/apache/iceberg/expressions/Literals.java#L444-L445

File: api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java (hunk 3 above)
Comment (ebyhr) [reply api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java:112]: Sure, added the test case.

File: api/src/test/java/org/apache/iceberg/types/TestConversions.java (hunk 2 above)
Comment (amogh-jahagirdar) [reply api/src/test/java/org/apache/iceberg/types/TestConversions.java:115]: Ah I see the comment on line 107/108. Could we update the `assertConversion` to instead test against 400000L and then remove the comment. At this point we are no longer having to pass in different values since we Literal.of(someLong).to(TimestampNanos) will always interpret someLong as nanoseconds.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: FWIW I still think this is correct but it's worth getting other's perspective on this since we are changing one of the assumptions of how value is interpreted when the type to convert to is nanoseconds.

CC @nastra @epgif @jacobmarble @rdblue thoughts?

Comment (Fokko) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: I think both before and after this change is correct. In Iceberg we had the assumption that _everything is in microseconds_. But this doesn't hold anymore now we have nano's. I do think the version after the change is more correct and more closely aligns with my expectations. If we can make sure that folks are not using this yet, I think this change is a good one 👍

Comment (Fokko) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: I had a chat with @rdblue who reviewed the PR that introduced this, and it is actually on purpose. Spark always passes in microseconds, changing this would break this assumption with Spark. So I think we have to revert this line. That said, I do think we need to check (and raise an error) when it overflows. Easiest way of doing this is by converting it to nano's, and convert is back to micro's and check if it still the same value.

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Thank you for sharing the context. What about other query engines? Actually, I found this issue when I was trying to support nanosecond precision in Trino Iceberg connector. As you may know, the max precision in Trino is picos (12).

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @Fokko Could you confirm the above question? The current implementation is weird from other query engine's perspectives.

Comment (Fokko) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Sorry, the example I gave was not the best one. Previous to V3, Iceberg did not have nanoseconds, so therefore we have the assumption that every long we see, is in microseconds. Otherwise, the following query would change:

```sql
CREATE TABLE tbl(ts timestamp);

-- We want to have all the events from the future
SELECT * FROM tbl WHERE ts > 1739288553127 -- this is interpreted as microseconds

-- We need more precision
ALTER TABLE tbl MODIFY COLUMN ts timestamp_ns

SELECT * FROM tbl WHERE ts > 1739288553127t -- this is interpreted as nanoseconds, changing the result
```

The trick here is to let Trino construct `TimestampNanoLiteral` and push that into the evaluator, instead of a plain `LongLiteral`.

Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Please forgive my delayed participation here. I'm not familiar with either Spark or Trino, but @epgif and I did author the nanoseconds PR.

Reviewing @rdblue's comment https://github.com/apache/iceberg/pull/9008#discussion_r1697460829 is where we switched away from `new TimestampNanoLiteral(value())`.

> we have the assumption that every long we see, is in microseconds

This is the core of the problem, right?

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Here is the problematic place in Trino: 
https://github.com/trinodb/trino/blob/be9ae2f2d61aeee03352c34326416ea7e7fe1354/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ExpressionConverter.java#L112-L121

```java
            for (Range range : orderedRanges) {
                if (range.isSingleValue()) {
                    icebergValues.add(convertTrinoValueToIceberg(type, range.getLowBoundedValue()));
                }
                else {
                    rangeExpressions.add(toIcebergExpression(columnName, range));
                }
            }
            Expression ranges = or(rangeExpressions);
            Expression values = icebergValues.isEmpty() ? alwaysFalse() : in(columnName, icebergValues);
```

The `convertTrinoValueToIceberg` method returns Long (not literal) for timestamp(9) type. The bottom `values` expression is created with Iceberg's `Expressions.in` method. 

> let Trino construct TimestampNanoLiteral and push that into the evaluator, instead of a plain LongLiteral.

TimestampNanoLiteral and other literals' code are package-private. Let me know if there is a method we can use from other repositories.

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @Fokko Could you confirm the above comment? This is a blocker of nanosecond support in Trino.

Comment (Fokko) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Hey @ebyhr sorry for the late reply. I wanted to check with some Java folks since I don't want to butcher the API. How about adding some factory methods?

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @Fokko Could you re-open this PR? I tried string literal approach. It works on data columns, but results in exception on partition columns. 

For instance, if we pass `ref(name="part_col") in ("2022-07-26T12:13:14.123456789")` to  `Scan#filter` method, the `Projections` line converts the expression to `ref(name="part_col") == 1658837594123456789` and then results in the following stacktrace.

https://github.com/apache/iceberg/blob/84741867b9220f47b8e9a83edfb1be14d40cbefc/core/src/main/java/org/apache/iceberg/ManifestGroup.java#L253-L255

```
java.lang.ArithmeticException: long overflow
	at java.base/java.lang.Math.multiplyExact(Math.java:1033)
	at org.apache.iceberg.util.DateTimeUtil.microsToNanos(DateTimeUtil.java:101)
	at org.apache.iceberg.expressions.Literals$TimestampLiteral.to(Literals.java:445)
	at org.apache.iceberg.expressions.Literals$LongLiteral.to(Literals.java:304)
	at org.apache.iceberg.expressions.UnboundPredicate.bindLiteralOperation(UnboundPredicate.java:171)
	at org.apache.iceberg.expressions.UnboundPredicate.bind(UnboundPredicate.java:122)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:159)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:118)
	at org.apache.iceberg.expressions.ExpressionVisitors.visit(ExpressionVisitors.java:347)
	at org.apache.iceberg.expressions.Binder.bind(Binder.java:60)
	at org.apache.iceberg.expressions.ManifestEvaluator.<init>(ManifestEvaluator.java:67)
	at org.apache.iceberg.expressions.ManifestEvaluator.forPartitionFilter(ManifestEvaluator.java:63)
	at org.apache.iceberg.ManifestGroup.lambda$entries$8(ManifestGroup.java:252)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.lambda$newMappingFunction$3(LocalLoadingCache.java:185)
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.lambda$computeIfAbsent$2(UnboundedLocalCache.java:297)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1713)
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.computeIfAbsent(UnboundedLocalCache.java:293)
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:112)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:60)
	at org.apache.iceberg.ManifestGroup.lambda$entries$9(ManifestGroup.java:275)
	at org.apache.iceberg.io.CloseableIterable$5.shouldKeep(CloseableIterable.java:156)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:66)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:49)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:64)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:49)
	at org.apache.iceberg.io.CloseableIterator$3.hasNext(CloseableIterator.java:91)
	at org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:45)
	at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.hasNext(CloseableIterable.java:264)
```

**Comment by ajantha-bhat** [comment]
We faced this at Dremio too : https://github.com/apache/iceberg/issues/13160

Looks forward to this fix.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @ebyhr for the exception in your last comment with string literal, I assume the partition column is a hour/day type of transformation on a timestamp_ns column? for line 255 (`Projections` line), what is the literal for the `dataFilter`?

I understand it was done for compatibility with Spark. But I feel the after change is more intuitive than before. long literal value should be assumed with the same precision as the timestamp field type. 

string literal can express different precisions in the string. I am wondering if Spark will behave correctly if the filter value is a string literal with nano precision. If it works correctly, can we ask Spark users not to use long literal for timestamp_ns?

Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @stevenzwu imo I think that the fix as it is, is good as is. I'm not entirely sure I get why we need to keep this behavior because the case where there's a chance for a correctness issue about is Spark microsecond values being interpreted as nanoseconds for a timestamp_nano type. However, spark doesn't even support this data type yet so this situation doesn't seem possible yet so it feels like the right thing to do is to invalidate this assumption

Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: long literals is implicit and some precision interpretation rule has to be applied. The good practice is probably using the timestamp functions to explicit describe the precision, e.g. Spark SQL
```
TIMESTAMP '2021-06-30T02:00:00.000Z'
timestamp_micros(1230219000123123)
```

Until Spark adds timestamp_nano data type and SQL function, maybe Iceberg timestamp_nano just doesn't need to work for Spark? Also Spark can't create an Iceberg table with timestamp_nano field. probably also reasonable for read to fail too.

File: api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java [hunk 4]
```diff
@@ -107,6 +108,33 @@ public void testTimestampMicrosToDateConversion() {
...
+    // verify round-trip between timestamp_ns and long
+    Literal<Long> timestampNano =
+        Literal.of("2017-11-16T14:31:08.000000001").to(Types.TimestampNanoType.withoutZone());
```
Comment (stevenzwu) [inline api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java:115]: nit: `.123456789` is easier to read that there are 9 digits for nano

I created a PR for this along with changes on naming. `timestampNano` and `longLiteral` doesn't fully capture what they represent.
https://github.com/ebyhr/iceberg/pull/1

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: > I understand it was done for compatibility with Spark

I think this is a little inaccurate. The cast from long to a microsecond timestamp was done for compatibility with Spark. But making the long to nanosecond timestamp cast use microseconds was done for consistency within the Iceberg API.

The problem is: how should the API interpret a value when we don't know what unit the value is in? Originally, we didn't convert long to timestamp because of this problem. There is no way for the API to distinguish between these cases:

```java
Expressions.lessThan("ts", System.currentTimeMillis())
```
```java
Expressions.lessThan("ts", System.currentTimeMillis() * MILLIS_TO_MICROS)
```

In order to support Spark, we decided that long values should be interpreted as microseconds. Then when adding a `timestamp(9)` type we followed that decision for consistency. What I mean by consistency is that **an expression behaves the same way regardless of binding type**.

Think of an expression that uses a string:

```java
Expressions.lessThan("ts", "2022-07-26T12:13:14.123") // ts is timestamp(6)
```
```java
Expressions.lessThan("ts", "2022-07-26T12:13:14.123") // ts is timestamp(9)
```

Both of those expressions select the same values when they are cast between micros and nanos. What I mean is that they both select based on the same point in time, regardless of the precision of the column the expression is eventually bound to. If you use that expression to filter two different tables with different `ts` precisions, you would get essentially the same results.

Then the question is whether you should get the same result when passing a long value instead:

```java
long boundaryValue = parseTimestampMicros("2022-07-26T12:13:14.123");
Expressions.lessThan("ts", boundaryValue) // ts is timestamp(6)
```
```java
Expressions.lessThan("ts", boundaryValue) // ts is timestamp(9)
```

I think that you should get the same behavior when using an identical expression, just like you would when binding a string timestamp to a different type. The alternative is surprising, at least to me: when selecting from two tables that differ only by the timestamp type (nanosecond values end in 000) the results of the same expression are completely different. That means in order to construct an expression, you need to know the actual precision of the column in some cases.

I think that the right solution is to add a public method to construct timestamp literals from nanos and micros. It also looks like we need to look into expression binding because the stack trace above is concerning.

File: api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java (hunk 4 above)
Comment (ebyhr) [reply api/src/test/java/org/apache/iceberg/expressions/TestTimestampLiteralConversions.java]: Thanks for your suggestion, updated.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: I think the current behavior is pretty surprising since you can imagine that if I come to Iceberg and I never use non-nano timestamps and I pass in a Long as a literal I probably would assume I'm writing a nano-precision long. 

I think it may make sense to just *not* support long -> nanoTimestamp in this transform at all and instead throw an error saying to use a TimestampNanoLiteral (which we would make.) Unfortunately we are past the time when we could make that change painlessly so it's probably not possible. 

What we can do is catch the overflow and throw a smarter message like "Literal.of(Long) assumes timestamp is given in microseconds and the value given overflows the timestamp container. Use Literals.timestampNanos(Long nanoseconds) to create a literal with nanosecond resolution"

Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: I like Russel's suggestion of converting overflow error to an exception with clear error message since we want to keep the current behavior of interpreting long literals consistently as micro. For nano timestamp, users just have to use string literal for now.

Maybe in a separate effort, we can think about how to expose `TimestampLiteral` and `TimestampNanoLiteral` to clients/engines. that is when these use cases can be supported.
```
ts < timestamp_micros(1230219000123123)
ts < timestamp_nano(1230219000123123)
```

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: >  I assume the partition column is a hour/day transformation on a timestamp_ns column?

No, it's a partition column without transformation. 

> For nano timestamp, users just have to use string literal for now.

This approach doesn't work well on partition columns. I will add a test to showcase the scenario.

Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: > No, it's a partition column without transformation.

Partition on a timestamp column (micro or nano) without transformations like hour, day? Identity partition on ts column doesn't seem right.

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: @stevenzwu I don't think we want to discuss the best practice for partitioning. Iceberg spec allows such a partition spec, right? It's just one of our test scenarios at Starburst, not a customer's table definition.

Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: sure. But at least the broken behavior using string literal with identity partition transformation on timestamp_ns column has minimal/no impact in practice, since it is not really a viable setup with production workload. the string literal problem can be tracked separately than the long literal issue.

Come back to this PR/issue, we need to decide on what to do with long literal for timestamp_ns. there are two options (1) derive/imply the precision (micro or nano) based the timestamp column type, which is the current approach this PR (2) always assume micro second precision for consistent interpretation behavior regardless of column timestamp precision, which is the current behavior and where Ryan coming from.

It seems that there is no consensus on adopting the behavior change of option 1. To move forward without changing current behavior and causing confusions to users, we can adopt these two steps
1) for the immediate step, fail the long literal (assumed micro precision) to nano ts with an explicit and more friendly error msg (suggested by Russel). For now, users should use the string literal with hour/day/week/month transformation on timestamp_nano partition column.
2) for the longer term, we can expose public methods for engines/clients to construct `TimestampLiteralNano` directly. Then long literal can used correctly with timestamp nano type, e.g. `ts < timestamp_nano(1230219000123123)`

Comment (stevenzwu) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: What about these two options of new public API for literal construction in `Expressions` class? With the new pubic constructor, would it be enough for engines to support nano timestamp?

1. `public static Literal<TimestampNanoLiteral> litTimestampNano(long value) `

With this API, clients/engines could construct literals like this when they know the long value contains timestamp in nano.
```
Literal<?> columnLiteral;
if (icebergType.typeId() == TypeID.TIMESTAMP_NANO && columnValue instanceof Long) {
  columnLiteral = Expressions.litNanoTimestamp(columnValue);
} else {
  columnLiteral = Expressions.lit(columnValue);
}
```

2. `public static <T> Literal<T> lit(T value, TypeID type)`

Basically move the above if-else inside this method, which assumes the long value for nano timestamp type has the precision of nano.

Comment (ebyhr) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: Adding such method is insufficient if I understand correctly. For instance, [the problematic place in Trino](https://github.com/apache/iceberg/pull/11775#discussion_r1951940306) depends on `Expressions#in(java.lang.String, T...)` taking values, not literals.

**Comment by rdblue** [comment]
Following up on the bug described in [this comment](https://github.com/apache/iceberg/pull/11775#discussion_r2074715304) here instead of in that thread because the thread discusses solutions to handling nanos, not just the overflow problem.

Here's a test that reproduces the problem:
```java
  @Test
  public void testStringToTimestampNanosLiteral() {
    Schema schema = new Schema(
        Types.NestedField.required(1, "id", Types.LongType.get()),
        Types.NestedField.optional(2, "ts", Types.TimestampNanoType.withoutZone()));

    PartitionSpec spec = PartitionSpec.builderFor(schema).identity("ts").build();

    Expression expr = Expressions.equal("ts", "2022-07-26T12:13:14.123456789");
    Expression projected = Projections.inclusive(spec).project(expr);

    Binder.bind(schema.asStruct(), projected);
  }
```

What's happening is the projection will first bind the original predicate because it needs a bound ID reference rather than a name reference. That ID is used to find the partition fields that can project the predicate. The binding process produces the expected `TimestampNanoLiteral(1658837594123456789L)`. The bug is in the `Identity` transform's projection code, which needs to produce a new predicate that is unbound and uses a reference for the partition field's name (the partition name does not have to match). When it [constructs the new unbound predicate](https://github.com/apache/iceberg/blob/e1e0a74/api/src/main/java/org/apache/iceberg/transforms/Identity.java#L152-L153), it passes the underlying value rather than the unchanged literal.

Updating that line to pass the literal instead of the value fixes the problem because it doesn't lose the context that the value was already a nanosecond timestamp value. I'll open a PR to fix this. Hopefully it can make it into 1.10.

**Comment by rdblue** [comment]
Opened #13746.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:303]: I just opened #13747 with fixes for the problems raised by this thread. It adds `millis`, `micros`, and `nanos` methods to `Expressions` that construct timestamp literals from long values. I also updated `Literals.from` so that these literals can be passed when creating `in` expressions (see the [test](https://github.com/apache/iceberg/pull/13747/files#diff-26e943160c86dac046e620ca3844a909ac22c3f1a06e63671b6e8b2484ab3fddR87)).

**Comment by rdblue** [comment]
This was addressed by #13746 and #13747. I'll close this.
